internal/sdk/crypto/crypto.go

```diff
@@ -0,0 +1,222 @@
+	}
+	key, err := parseRSAPublicKey(res)
+	if err != nil {
+		return "", err
+	}
+	v, err := encryptWithRSAPublicKey(value, key)
+	if err != nil {
+		return "", err
+	}
+
+	return v, nil
+}
+
+func (c *Crypto) fetchPublicKey(ctx context.Context, tlsCert tls.Certificate) (pem []byte, err error) {
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs: c.RootCAs,
+				Certificates: []tls.Certificate{
+					tlsCert,
+				},
+			},
+		},
+		Timeout: time.Minute,
+	}
+	url := fmt.Sprintf("%s/key", c.URL)
+	body, err := sdkHttp.Do(ctx, httpClient, http.MethodGet, url, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	if _, err := decodeRSAPublicKey(body); err != nil {
+		return nil, fmt.Errorf("GET %s: parse body %q: %v", url, string(body), err)
+	}
+	return body, nil
+}
+
+func split(data []byte) [][]byte {
+	var r [][]byte
+	for offset, l := 0, len(data); offset < l; {
+		n, token := scanPEMs(data[offset:], true)
+		if token != nil {
+			r = append(r, token)
+		}
+		if n == 0 {
+			panic("pem: split: advance is 0")
+		}
+		offset += n
+	}
+	return r
+}
+
+func scanPEMs(data []byte, atEOF bool) (advance int, token []byte) {
+	if atEOF && len(data) == 0 {
+		return 0, nil
+	}
+	var offset int
+	for {
+		d := data[offset:]
+		if i := bytes.IndexByte(d, '\n'); i != -1 {
+			// Check if the line following \n starts with "-----END ".
+			if bytes.HasPrefix(d[i+1:], pemEndPrefix) {
+				// Find the end of line.
+				if j := bytes.IndexByte(d[i+1:], '\n'); j != -1 {
+					// Check if line ends with "-----".
+					if bytes.HasSuffix(d[:i+1+j], pemEndSuffix) {
+						return offset + i + 1 + j + 1,
+							bytes.TrimSpace(data[:offset+i+1+j])
+					} else {
+						// Try next line.
+						offset += i + 1 + j + 1
+						continue
+					}
+				}
+				// Break.
+			} else {
+				// Try next line.
+				offset += i + 1
+				continue
+			}
+		}
+		break
+	}
+	// If we're at EOF, we have a final, non-terminated line. Return it.
+	if atEOF {
+		return len(data), bytes.TrimSpace(data)
+	}
+	// Request more data.
+	return 0, nil
+}
+
+func x509KeyPairWithLeaf(certPEMBlock, keyPEMBlock []byte) (tls.Certificate, error) {
+	tlsCert, err := tls.X509KeyPair(certPEMBlock, keyPEMBlock)
+	if err != nil {
+		return tls.Certificate{}, err
+	}
+	leaf, err := x509.ParseCertificate(tlsCert.Certificate[0])
+	if err != nil {
+		return tls.Certificate{}, err
+	}
+	tlsCert.Leaf = leaf
+	return tlsCert, nil
+}
+
+func decodeRSAPublicKey(data []byte) (*pem.Block, error) {
+	return Decode(data, "RSA PUBLIC KEY")
+}
+
+func parseRSAPublicKey(data []byte) (*rsa.PublicKey, error) {
+	block, err := Decode(data, "RSA PUBLIC KEY")
+	if err != nil {
+		return nil, err
+	}
+	return x509.ParsePKCS1PublicKey(block.Bytes)
+}
+
+func encryptWithRSAPublicKey(token string, pub *rsa.PublicKey) (string, error) {
+	hash := sha256.New()
+	ciphertext, err := rsa.EncryptOAEP(hash, rand.Reader, pub, []byte(token), nil)
```

Comment:
From [OAEP RFC](https://www.rfc-editor.org/rfc/rfc8017#section-7.1.1): 

```
    M        message to be encrypted, an octet string of length mLen,
               where mLen <= k - 2hLen - 2
```
 
Consider either adding a check for max length (`maxLength := pub.Size() - 2*has.Size() -2`) or (perhaps preferably) using hybrid encryption (encrypt token with secure symmetric key, then encrypt symmetric key with RSA).

Reply:
If we change this here I guess we need to apply the same changes when we encrypt/decrypt in the control plane, right?

Reply:
This validations is internally applied at `rsa.EncryptOAEP`